### PR TITLE
ARROW-252: Add implementation guidelines to the documentation

### PR DESCRIPTION
**Pull request by julienledem**

### CODE REVIEW

**Comment by wesm** [comment]
Looks good to me. Would be great for someone else to take a look too

File: format/Guidelines.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+### If only producing (and not consuming) arrow vectors.
+any subset of the vector spec and the corresponding metadata can be implemented
+
+### If consuming and producing vectors
+There is a minimal subset of vectors to be supported.
+Production of a subset of vectors and their corresponding metadata is always fine.
+Consumption of vectors should at least convert the unsupported input vectors to the supported subset (for example Timestamp.millis to timestamp.micros or int32 to int64)
+
+## Extensibility
+An execution engine implementor can also extend their memory representation with their own vectors internally as long as they are never exposed. Before sending data to another system expecting Arrow data these custom vectors should be converted to a type that exist in the Arrow spec.
```
Comment (emkornfield) [inline format/Guidelines.md]: Mention whether custom metadata are in/out of scope for V1 (or ever).  I would initially vote for out of scope to keep things simple.

File: format/Guidelines.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Implementation guidelines
+
+An execution engine (or framework, or UDF executor, or storage engine, etc) can implements only a subset of the Arrow spec and/or extend it given the following constraints:
+
+## Implementing a subset the spec
+### If only producing (and not consuming) arrow vectors.
+any subset of the vector spec and the corresponding metadata can be implemented
```
Comment (emkornfield) [inline format/Guidelines.md]: minor.  Capitalize "any", add period at the end.

Comment (julienledem) [inline format/Guidelines.md]: done

File: format/Guidelines.md (hunk 1 above)
Comment (julienledem) [inline format/Guidelines.md]: done

**Pull request closed without merging**